### PR TITLE
feat: full beam filter consistency for H1-H5 in songset constructor

### PR DESCRIPTION
**Pull request by mhuang74**
## Summary

Generalizes the H2/H3 endcap fix to **every relaxable rule** (H1, H4, H5) whose relaxation flag is wired through `RunConfig` but never consulted by the beam's sequence builder. Previously, `_sequences` filtered only by phase, leaving H1/H4/H5 to `validate()` after the beam converged — so relax flags couldn't influence which sequences were generated, and diagnostics reported rejection counts that didn't match the beam's actual escape tiers.

This PR implements spec `songset-constructor-beam-in-sequence-filtering-v3.md`, which subsumes v1 (H2/H3 endcap filtering) and v2 (deferred items) entirely.

**Before:** `--relax-h3-bpm 110` produced `candidates=0` because the beam converged on a 161 BPM closer that `validate()` rejected, and the relax flag had no way to influence which closers were proposed.

**After:** The beam filters at the source — endcap tempo limits (H2/H3), opener phase set (H1), per-pair BPM delta (H4), and per-pair circle-of-fifths distance (H5) are all consulted during sequence generation, using the runtime config. Relax flags now have teeth.

### `config.py` — new H4/H5 limit properties + CLI plumbing
- Added `relax_h4`, `relax_h5` boolean fields and `relax_h4_bpm`, `relax_h5_cfd` optional numeric override fields
- Added `h4_limit` property (20 strict / 25 relaxed / override) and `h5_limit` property (2 strict / 3 relaxed / override), mirroring existing `closing_limit`/`opening_floor`
- Added validation guards for negative `--relax-h4-bpm` / `--relax-h5-cfd`
- Updated `to_dict()` to round-trip all 4 new fields (guardrail for `compact_config` roundtrip in `beam.py`)

### `beam.py` — full source filtering + compute_fan_out parameterization
- **`compute_fan_out`** signature now takes `config: RunConfig`; hard-coded `cfd <= 2 and bpm_delta <= 20` replaced with `cfd <= config.h5_limit and bpm_delta <= config.h4_limit`
- **`_sequences`** signature changed from `(pool, songs, width)` to `(pool, config, matrix, width)`; now filters at the source:
  - **Opener slot (H1 + H2):** accepts `phase in {1, 2}` when `config.relax_h1`, else `phase == 1`; enforces `tempo_bpm >= config.opening_floor`
  - **Closer slot (H3):** enforces `phase in {4, 5}` and `tempo_bpm <= config.closing_limit`
  - **Per-pair expansion (H4):** looks up `transition = matrix.get((left, right))`; computes `bpm_delta`; applies `allowed = config.h4_limit` if transition has crossfade/gap, else `min(15, config.h4_limit)`; rejects if `bpm_delta > allowed`
  - **Per-pair expansion (H5):** computes `distance = transition.cfd`; applies `shifted_ok` carve-out when `transition.suggested_key_shift != 0`; rejects if `distance > config.h5_limit and not shifted_ok`
- **Auto-escalation tiers** now build `relaxed_config` with `relax_h4=True, relax_h5=True` so the source filter widens to match the relaxation; each relaxed tier re-calls `compute_fan_out(pool, matrix, relaxed_config)` and re-sorts so dead-end flags reflect the widened limits (prevents over-pruning in relaxed tiers)
- **First tier `validate()`** now propagates `relax_h1=config.relax_h1, relax_h4=config.relax_h4, relax_h5=config.relax_h5` (previously called with no relax kwargs, causing H1 default-True to be ignored)

### `diagnostics.py` — honest tier reporting
- `beam_diagnostics` now imports and uses `beam._candidate_sort_key` (kills sort-key drift between diagnostics and beam)
- `hard_rule_rejection_counts` gains optional `relax_kwargs: dict | None = None` parameter forwarded to `validate()`
- `beam_diagnostics` produces both:
  - `hard_rule_rejections` — strict-mode rejection counts (current behavior)
  - `relaxed_tier_rejections` — rejection counts under the `relaxed_H4_H5` escape tier, so users see how the beam would have actually escaped
- `_proposal_for_diagnostics` now applies transition values (crossfade, gap_beats, key_shift) to proposal items, matching `_proposal_for_sequence` in beam.py — ensures `validate()` in diagnostics sees the same crossfade/gap settings the beam used
- `diagnostic_lines` emits a second line when `relaxed_tier_rejections` differs from strict: `beam relaxed-tier fallback would still reject N/M sequences: H4=X, H5=Y`

### `graph/nodes.py` — LangGraph relaxation parity
- `validate_score` (`nodes.py:199-204`) now propagates `relax_h4=config.relax_h4` and `relax_h5=config.relax_h5` alongside `relax_h1` — fixes the agentic/deterministic beam divergence where the LangGraph path disagreed with the beam fallback on what counts as valid
- `build_transition_matrix` (`nodes.py:99`) passes `state["config"]` to `compute_fan_out`

### `cli.py` — new user-facing CLI flags
- Added `--relax-h4/--no-relax-h4`, `--relax-h5/--no-relax-h5`, `--relax-h4-bpm` (min=0), `--relax-h5-cfd` (min=0)
- All wired through to `RunConfig` constructor and surface in `--help`

### Tests — 22 new tests across 4 files

**`test_songset_constructor_rules.py` (14 new tests):**
- v1 H2/H3 endcap: `test_beam_filters_closer_by_h3_ceiling`, `test_beam_filters_opener_by_h2_floor`, `test_relax_h3_unblocks_when_only_high_bpm_closer_matches_preceding`, `test_diagnostics_beam_sequences_uses_config_ceiling`
- v3 H1: `test_relax_h1_opener_accepts_phase_2` (phase-2 opener accepted when `relax_h1=True`, rejected when `False`)
- v3 H4: `test_beam_rejects_h4_violating_middle_pair` (strict 0 / relaxed >0 via `_sequences`), `test_beam_h4_honors_crossfade_branch` (pair with `crossfade_duration_seconds > 0` and `bpm_delta=18` survives strict `_sequences`)
- v3 H5: `test_beam_rejects_h5_violating_pair` (strict 0 / relaxed >0), `test_beam_h5_honors_suggested_key_shift` (pair with CFD=4 and `suggested_key_shift != 0` survives), `test_beam_h5_shifted_ok_after_proposal_applies_shift` (end-to-end invariant: `_proposal_for_sequence` applies the shift, `validate()` sees it)
- v3 diagnostics: `test_diagnostics_relaxed_tier_report_present`, `test_diagnostics_uses_beam_sort_key` (monkeypatch guard against sort-key drift regression)
- v3 config: `test_to_dict_preserves_relax_h4_h5`, `test_compute_fan_out_uses_config_limits`

**`test_songset_constructor_config.py` (3 new tests):**
- `test_config_h4_limit_property`, `test_config_h5_limit_property`, `test_config_relax_h4_h5_negative_raises_value_error`

**`test_songset_constructor_graph.py` (1 new test):**
- `test_validate_score_propagates_relax_h4_h5` — mocks `validate` to capture kwargs, asserts `relax_h4=True` and `relax_h5=True` are propagated

**`test_songset_constructor_cli.py` (2 new tests):**
- `test_cli_relax_h4_flag_args_accepted`, `test_cli_relax_h5_flag_args_accepted`

## Design decisions

- **Filter at the source, don't change sort key weights.** The BPM-delta and CFD *weights* in the sort key are unchanged; the filters just prune candidates that violate the relaxable ceilings at each expansion step.
- **Auto-escalation tier order unchanged:** strict → relaxed_H4_H5 → relaxed_H2_H3 → relaxed_H1+H2+H3+H4+H5. This prioritizes keeping hard musical limits intact for as long as possible.
- **Strict compatibility preserved.** When no relax flags are set, behavior on previously-passing pools is unchanged: candidates that would have failed `validate()` are now removed up-front (marginal cost: dropped candidates that would never have passed anyway).
- **H4 crossfade branch consults the transition matrix per-pair** inside `_sequences`, matching `validate()`'s `allowed = h4_limit if (crossfade > 0 or gap_beats > 4) else min(15, h4_limit)` logic.
- **H5 `shifted_ok` carve-out** treats every `suggested_key_shift != 0` pair as "shifted ok" in `_sequences` because `_proposal_for_sequence` applies the suggestion later — so by the time `validate()` runs, the shift is in place and the carving is consistent end-to-end.

## Testing notes

```bash
# All 47 songset_constructor tests pass
PYTHONPATH=lab/poc-scripts uv run --project lab/poc-scripts \
  --extra songset_constructor --extra test pytest \
  lab/poc-scripts/tests/test_songset_constructor_rules.py \
  lab/poc-scripts/tests/test_songset_constructor_config.py \
  lab/poc-scripts/tests/test_songset_constructor_cli.py \
  lab/poc-scripts/tests/test_songset_constructor_graph.py -v
```

The 9 failures in `test_eval_lrc.py` are pre-existing (missing `pypinyin` module) and unrelated to this PR.

## Edge cases / risks

- **Per-pair H4 filter is stricter than the old "validate after the fact"** for pools with sparse matrices (many missing transitions fall back to `min(15, h4_limit)` and CFD=6 sentinel). The `relaxed_H4_H5` tier remains the escape hatch.
- **`compact_config` round-trip** (`beam.py` 4-song fallback) now preserves `relax_h4`/`relax_h5`/`relax_h4_bpm`/`relax_h5_cfd` via `to_dict()`. Covered by `test_to_dict_preserves_relax_h4_h5`.
- **Diagnostics overhead** — `validate()` runs twice per sequence in `beam_diagnostics` (strict + relaxed). Beam width is small (8) and this is a diagnostic path; acceptable.

### CODE REVIEW

File: lab/poc-scripts/poc/songset_constructor/cli.py [hunk 1]
```diff
@@ -313,6 +313,10 @@ def construct(
     relax_h2_bpm: Annotated[int | None, typer.Option("--relax-h2-bpm", min=0)] = None,
     relax_h1: Annotated[bool, typer.Option("--relax-h1/--no-relax-h1")] = True,
     auto_relax: Annotated[bool, typer.Option("--auto-relax/--no-auto-relax")] = True,
+    relax_h4: Annotated[bool, typer.Option("--relax-h4/--no-relax-h4")] = False,
+    relax_h5: Annotated[bool, typer.Option("--relax-h5/--no-relax-h5")] = False,
+    relax_h4_bpm: Annotated[int | None, typer.Option("--relax-h4-bpm", min=0)] = None,
+    relax_h5_cfd: Annotated[int | None, typer.Option("--relax-h5-cfd", min=0)] = None,
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/poc/songset_constructor/cli.py:330]: Fixed in 374155c. `validate()` now uses `config.h4_limit` and `config.h5_limit` instead of the hardcoded `25 if relax_h4 else 20` / `3 if relax_h5 else 2` ceilings. Since `RunConfig.h4_limit` already incorporates both `relax_h4` and `relax_h4_bpm` (and likewise for h5), the custom CLI overrides now propagate end-to-end from beam source filter through validation.

File: lab/poc-scripts/poc/songset_constructor/rules/beam.py [hunk 2]
```diff
@@ -28,7 +29,7 @@ def compute_fan_out(
             if candidate.recording_hash_prefix == other.recording_hash_prefix:
                 continue
             transition = matrix.get((candidate.recording_hash_prefix, other.recording_hash_prefix))
-            if transition and transition.cfd <= 2 and transition.bpm_delta <= 20:
+            if transition and transition.cfd <= config.h5_limit and transition.bpm_delta <= config.h4_limit:
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/poc/songset_constructor/rules/beam.py]: Fixed in 374155c. `compute_fan_out` now counts an edge as viable when `transition.bpm_delta <= config.h4_limit AND (transition.cfd <= config.h5_limit OR transition.suggested_key_shift != 0)`. This matches the transposition exemption that `_sequences()` already applies at line 106, so songs whose only viable outgoing transitions rely on suggested transposition are no longer incorrectly marked as `is_dead_end=True`.

The test `test_beam_rejects_h5_violating_pair` was updated to zero out `suggested_key_shift` on H5-violating edges, so it correctly tests the actual intent: H5 violations without transposition are rejected.

File: lab/poc-scripts/poc/songset_constructor/artifacts/writer.py [hunk 3]
```diff
@@ -4,8 +4,10 @@
 
 import csv
 import json
+from collections import Counter
 from datetime import UTC, datetime
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/poc/songset_constructor/artifacts/writer.py:8]: No change needed. This project requires Python >=3.11 (`pyproject.toml` line 5: `requires-python = ">=3.11"`), and `from datetime import UTC` is already the established pattern in `config.py:7`. The `UTC` import is safe for the supported runtime.

File: lab/poc-scripts/poc/songset_constructor/artifacts/writer.py [hunk 4]
```diff
@@ -133,3 +140,360 @@ def write_pool_csv(path: Path, pool: list[SongCandidate]) -> None:
...
+def build_review_report(
+    *,
+    config: RunConfig,
+    proposals: list[SongsetProposal],
+    pool: list[SongCandidate],
+    trace: list[dict],
+    chat: Any | None = None,
+    fallback_note: str | None = None,
+) -> str:
+    generated_at = datetime.now(UTC).isoformat()
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/poc/songset_constructor/artifacts/writer.py:154]: No change needed. Same as above — the project targets Python >=3.11 (`pyproject.toml`), and `datetime.now(UTC)` is already used in `config.py:19,94`. Keeping the import consistent across modules.

File: lab/poc-scripts/poc/songset_constructor/rules/beam.py [hunk 5]
```diff
@@ -58,14 +64,47 @@ def _sequences(pool: list[SongCandidate], songs: int, width: int = 8) -> Iterabl
...
+                        transition.bpm_delta
+                        if transition
+                        else abs((candidate.tempo_bpm or 0) - (left.tempo_bpm or 0))
+                    )
+                    allowed = (
+                        config.h4_limit
+                        if transition
+                        and (transition.crossfade_duration_seconds > 0 or transition.gap_beats > 4)
+                        else min(15, config.h4_limit)
+                    )
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/poc/songset_constructor/rules/beam.py:117]: No change needed. In this codebase, `TransitionCandidate.crossfade_duration_seconds` and `gap_beats` are defined as required `float` fields (models.py:41-42), not `float | None`. Pydantic enforces them as non-`None`, so the `> 0` / `> 4` comparisons cannot raise `TypeError`. Additionally, the `transition` variable is already guarded by `if transition and ...` before field access.

File: lab/poc-scripts/diagnose_closers.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                            continue
+                    if beam and candidate.phase < beam[-1].phase - 1:
+                        continue
+                    if candidate.is_dead_end and position != len(target):
+                        continue
+                    if beam:
+                        left = beam[-1]
+                        transition = matrix.get((left.recording_hash_prefix, candidate.recording_hash_prefix))
+                        bpm_delta = transition.bpm_delta if transition else abs((candidate.tempo_bpm or 0) - (left.tempo_bpm or 0))
+                        allowed = cfg.h4_limit if (transition and (transition.crossfade_duration_seconds > 0 or transition.gap_beats > 4)) else min(15, cfg.h4_limit)
```
[the remark this answers is not in the document]
Comment (mhuang74) [reply lab/poc-scripts/diagnose_closers.py:162]: No change needed. Same reasoning as the beam.py comment above: `TransitionCandidate.crossfade_duration_seconds` and `gap_beats` are required `float` fields (models.py:41-42), non-nullable by Pydantic schema. The `TypeError` cannot occur.
